### ISSUE TITLE
use application classloader instead of system classloader

### ISSUE THREAD

**Issue by del-flexe** [johncarl81/parceler#196]
If I declare a `@Parcel` having a `android.os.Parcelable` member:

```java
@Parcel
class MyParcel<T extends Parcelable> {
  public final T item;

  @ParcelCreator
  public MyParcel(final T item) {
    this.item = item
  }
}
```

The generated code in the implementation of `Parcelable#read` attempts to deserialize the parcelable explicitly using the classloader of `android.os.Parcelable` which is **not** the application's classloader:

```java
android.os.Parcelable parcelable$$0 = ((android.os.Parcelable) parcel$$3 .readParcelable(android.os.Parcelable.class.getClassLoader()));
```

This means that Parceler allows successfully writing objects to a parcel for the type, but not reading them out as the application's class cannot be found during deserialization:

```
05-19 14:27:44.144  6772  6772 E AndroidRuntime: Caused by: android.os.BadParcelableException: ClassNotFoundException when unmarshalling: MyParcel$$Parcelable
05-19 14:27:44.144  6772  6772 E AndroidRuntime:        at android.os.Parcel.readParcelableCreator(Parcel.java:2432)
05-19 14:27:44.144  6772  6772 E AndroidRuntime:        at android.os.Parcel.readParcelable(Parcel.java:2358)
05-19 14:27:44.144  6772  6772 E AndroidRuntime:        at MyParcel$$Parcelable.read(...$Parcelable.java:76)
05-19 14:27:44.144  6772  6772 E AndroidRuntime:        at MyParcel$$Parcelable.read(...$Parcelable.java:84)
```

Manually setting the classloader before unwrapping in the activity has no effect, and neither does the "wrap the parcelable with a bundle" technique described here:

https://code.google.com/p/android/issues/detail?id=6822

I believe the problem code is here, but do not currently have a suggested mechanism for allowing use of the application classloader:

https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/generator/ParcelableReadWriteGenerator.java#L32

Can functionality be changed to allow selecting the classloader, or defaulting to the application classloader?

**Comment by johncarl81** [comment]
Shoot, it should probably use the generated `Parcelable.class.getClassLoader()` instead of the `returnJClassRef` (in this case `Parcelable`) classloader.  So that would change your example line to the following:
```
((android.os.Parcelable) parcel$$3 .readParcelable(MyParcel$$Parcelable.class.getClassLoader()));
```

**Comment by johncarl81** [comment]
or we could just give `null` to specify the default class loader. https://developer.android.com/reference/android/os/Parcel.html#readParcelable(java.lang.ClassLoader)

**Comment by del-flexe** [comment]
if it defaults to the application classloader if run in an activity, that would suit me just fine!

**Comment by johncarl81** [comment]
Here's the fix if you'd like to give it a try: https://github.com/johncarl81/parceler/pull/197
By the way, this was a recent change by me, not sure why I went this direction honestly: b222e7da1ce6809aeb80c3f40d8dd64e5467e509

**Comment by del-flexe** [comment]
sorry ... I don't see any info on building the project ... has the snapshot JAR been published somewhere?

**Comment by johncarl81** [comment]
here's a previous tutorial on building: https://github.com/johncarl81/parceler/issues/193#issuecomment-219593214
I guess this sort of thing could be added to the README.  Generally it's just `mvn clean install`.

**Comment by del-flexe** [comment]
looks like `mvn package` downloads JRuby?  o_O

```
Downloaded: http://repo.maven.apache.org/maven2/org/jruby/jruby-complete/1.7.16.1/jruby-complete-1.7.16.1.jar (21604 KB at 2899.8 KB/sec)
```

**Comment by johncarl81** [comment]
yup, we use Asciidoclet for javadoc comments.  If you want to avoid that you can add the skip javadoc option:
`mvn clean install -Djavadoc.skip=true`

**Comment by del-flexe** [comment]
doing an 'mvn install:install-file' with the `parceler` and `parceler-api` artifacts puts them in my `.m2` cache under the groups of the same name, but the _parceler-api-1.1.5-SNAPSHOT.pom_ references the `parceler` JAR in a different group: _parceler-project_

**Comment by johncarl81** [comment]
why are you running `install:install-file`?  `mvn install` should be exactly what you need.

**Comment by del-flexe** [comment]
built on different box ... looks like it ignored the artifact ID during the install.  In theory reading from the POM during the install should do the same thing

**Comment by del-flexe** [comment]
the patch works quite well =)

**Comment by johncarl81** [comment]
This is released under 1.1.5

**Comment by del-flexe** [comment]
awesome - thanks!
